### ISSUE TITLE
Auto-close hover (and possibly other) window

### ISSUE THREAD

**Issue by mandreyel** [natebosch/vim-lsc#97]
Hey!

Thanks for the plugin, it seems great so far--wonder why it hasn't gotten more attention!

I got a possibly stupid question (I'm unfortunately not versed in VimL, yet): is it currently possible to make the preview window provided by hover auto close after, say navigating away from the field whose information is being displayed? Currently I have to switch to that window and close it with :q, unless I've grossly misunderstood something.

Thanks in advance!

**Comment by natebosch** [comment]
`<C-w>z` (or `<C-w><C-z>` will close the preview window. See `:help preview-window`.

It would also be possible to do something like auto close the preview window on any movement but that's not something I'd like to put in to this plugin.

If you language server supports the `textDocument/documentHighlights` and you're seeing highlights of references to the field under the cursor we could also add an autocommand for when the highlighted field changes. I don't see that having wide appeal so I won't add it unless you think it'll be useful for the specific reason.

**Comment by mandreyel** [comment]
Oh, I didn't know about that command. Learn something new (particularly in Vim) every day!

I would say this is still not super ergonomic. What do think about posting the hover information as a status message? Or is there a rationale I'm not aware of for putting it in the preview window? Although I assume you've chosen to do this because there may not be enough space in the status line for the occasional longer hover information?

**Comment by prabirshrestha** [comment]
I would just put it in the doc since most vim users aren't' familiar with the bindings. That is what I did for vim-lsp.

Some people prefer it to be focused by default with q to quit, some prefer it not be focused because they can move around and see docs. Impossible to have a default that satisfies everyone.

**Comment by mandreyel** [comment]
@prabirshrestha Yes, that is a valid point, it's definitely difficult to pin down UX choices that please everyone.

In light of this, I think customization can help, so everyone can tailor the UX to fit their needs. But I understand if this is too high an undertaking for possibly not that much benefit.

I hope it's alright to ask a question unrelated to the issue: may I ask why you guys are not joining forces on a single vim LSP plugin? It seems like there is a lot of duplicate effort in this space (add to that LanguageClient-neovim). Granted, I'm not too familiar with either of your implementations and their design goals, so my apologies if there is an obvious difference between them that sets them apart.

**Comment by natebosch** [comment]
> What do think about posting the hover information as a status message?

That isn't very feasible in vim: https://github.com/vim/vim/issues/2463

> is there a rationale I'm not aware of for putting it in the preview window?

The typical vim approach would be to stick it in some window that isn't temporary like an `echo` would be. The preview window is a nicer experience than an ordinary buffer because of the shortcuts around it.

I could give you an autocommand that you could use as a handle to know when to set up a _separate_ autocommand on cursor move or something. I don't think I'd like to build it in to the plugin since it doesn't feel very vimmy to me, I think it might be better to help users get `<c-w><c-z>` into their muscle memory.

> I would just put it in the doc since most vim users aren't' familiar with the bindings.

Good call, I'll add a mention to the docs.

> may I ask why you guys are not joining forces on a single vim LSP plugin? It seems like there is a lot of duplicate effort in this space

FWIW I think it's good to have multiple options.
